keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,282 @@
+   3. `-m physdev --physdev-is-in` (for kubenet if we don’t want to depend on node podCIDR)
+
+Note the above are equivalent definitions, when considering only pod traffic originating from within the node.
+
+For the last use case, in iptables and ipvs, the proposal is to drop the reference to the cluster CIDR.
+
+The reasoning behind why this works are as follows.
+
+### iptables - masquerade off cluster traffic to services by node IP
+
+The [rule here currently](
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L887-L894
+) looks as follows
+
+```go
+// This masquerades off-cluster traffic to a service VIP.  The idea
+// is that you can establish a static route for your Service range,
+// routing to any node, and that node will bridge into the Service
+// for you.  Since that might bounce off-node, we masquerade here.
+// If/when we support "Local" policy for VIPs, we should update this.
+writeLine(proxier.natRules, append(args, "! -s", proxier.clusterCIDR, "-j", string(KubeMarkMasqChain))...)
+```
+
+The logic is that if the source IP is not part of the cluster CIDR range,
+then it must have originated from outside the cluster. Hence we add a rule to masquerade by
+the node IP so that we can send traffic to any pod within the cluster.
+
+One key insight when thinking about this data path though is the fact that the iptable rules run
+at _every_ node boundary. So when a pod sends a traffic to a service IP, it gets translated to
+one of the node IPs _before_ it leaves the node at the node boundary. So it's highly unlikely to 
+receive traffic at a node, whose destination is the service cluster IP, that is initiated by pods
+within the cluster, but not scheduled within that node.
+
+Going by the above reasoning, if we receive traffic whose source is not within the node generated pod traffic,we can say with very high confidence that the traffic originated from outside the cluster. This would be the simplest change with respect to re-writing the rule without any assumptions on how the pod networking is setup.
+
+### iptables - redirecting pod traffic to external loadbalancer VIP to cluster IP
+
+The [rule here currently](
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L1248-L1260
+) looks as follows
+
+```go
+// First rule in the chain redirects all pod -> external VIP traffic to the
+// Service's ClusterIP instead. This happens whether or not we have local
+// endpoints; only if clusterCIDR is specified
+if len(proxier.clusterCIDR) > 0 {
+  args = append(args[:0],
+    "-A", string(svcXlbChain),
+    "-m", "comment", "--comment",
+    `"Redirect pods trying to reach external loadbalancer VIP to clusterIP"`,
+    "-s", proxier.clusterCIDR,
+    "-j", string(svcChain),
+  )
+  writeLine(proxier.natRules, args...)
+}
+```
+
+The logic here is that if the source IP is part of cluster CIDR and we detect that is being
```

Comment:
I think this goes back to the comment on line 131. As long as we are talking about traffic "destined to services", it kind of seems to make sense to track the loadbalancer IP same as cluster IP when originating traffic from within the cluster?

Reply:
removing this will mean that for LB IPs there will be a bounce out of the cluster to the LB and back, which we don't have today. Other than that, I think it should continue to work? Based on how the cluster network is setup (say island mode), this might mean that as the traffic leaves the clusters, it first get SNAT by the node boundary and then goes to the LB and so we might end up with one extra conntrack too. So receiving backend pods can potentially see a change in source IP without this rule as against current behavior?

Reply:
I realized this would also be true when the load balancer IP is public IP (atleast in GCP) as all public outbound traffic is masqueraded by the node boundary before leaving the cluster. With the current rule, we only do a single DNAT. Without the rule it would be a SNAT (leaving the cluster) and DNAT (coming back in). So two conntrack entries instead of one.

Reply:
Should we couple the tweaking of LB IP "takeover" use case as part of this KEP or track it independently? If latter, should this be resolved?

Reply:
I think we should investigate getting rid of this but that is a separate issue - this is just adapting the existing faulty logic.